engine/src/execution_graph/logic_controllers/taskflow/kernel.cpp

```diff
@@ -9,5 +9,149 @@ std::pair<bool, uint64_t> kernel::get_estimated_output_num_rows(){
     return this->query_graph->get_estimated_input_rows_to_kernel(this->kernel_id);
 }
 
+void kernel::process(std::vector<std::unique_ptr<ral::cache::CacheData > > & inputs,
+		std::shared_ptr<ral::cache::CacheMachine> output,
+		cudaStream_t stream,
+        std::string kernel_process_name = ""){
+    std::vector< std::unique_ptr<ral::frame::BlazingTable> > input_gpu;
+
+    
+    if (this->has_limit_ && output->get_num_rows_added() >= this->limit_rows_) {
+  //      return;
+    }
+
+    for(auto & input : inputs){
+        try{
+            //if its in gpu this wont fail
+            //if its cpu and it fails the buffers arent deleted
+            //if its disk and fails the file isnt deleted
+            //so this should be safe
+            input_gpu.push_back(std::move(input->decache()));
+
+        }catch(std::exception e){
+            throw e;
+        }
+        
+
+    }
+
+    try{
+       do_process(std::move(input_gpu),output,stream, kernel_process_name);
+    }catch(std::exception e){
+        //remake inputs here
+        int i = 0;
+        for(auto & input : inputs){
+            if (input->get_type() == ral::cache::CacheDataType::GPU || input->get_type() == ral::cache::CacheDataType::GPU_METADATA){
+                //this was a gpu cachedata so now its not valid
+                static_cast<ral::cache::GPUCacheData *>(input.get())->set_data(std::move(input_gpu[i]));                 
+            }
+            i++;
+        }
+        throw;
+    }
+ 
+}
+
+void kernel::add_task(size_t task_id){
```

Comment:
I think that the `kernel::add_task` should take in as parameters:
`std::vector<std::unique_ptr<ral::cache::CacheData > > inputs, std::shared_ptr<ral::cache::CacheMachine> output, std::string kernel_process_name`

And it should be the one that calls `executor::add_task`. Then it will have the `kernel *` to pass it in and all the kernel run functions could be even cleaner, since they wont have this weird nested `add_task`  calls.  
Also, we can make it so that if you dont pass in the parameter `kernel_process_name`, we should have every kernel have its own function `get_kernel_name()` which should feed `kernel_process_name`. That way we can have the option of providing a custom `kernel_process_name` but most of the time it will be selected by the kernel itself. Plus i think it will be useful for the kernel to know its own name